Added timeout for running milestone service

aileen opened the pull request:
no issue

- Instead of running milestone service directly on boot, set a random timeout of 0-4 days to run after boot
- Updated tests
- Service is still behind a beta flag

aileen committed:
Added timeout for running milestone service

no issue

- Instead of running milestone service directly on boot, set a random timout of 1-4 days to run after boot
- Updated tests
- Service is still behing a beta flag

aileen commented on ghost/core/test/e2e-server/services/milestones.test.js:
```diff
@@ -206,29 +225,42 @@ describe('Milestones Service', function () {
+        const result = await resultPromise;
         assert(result === undefined);
+
+        assert(loggingStub.called);
```
No idea why, but the stub returns a call count much MUCH higher than I anticipate and also am receiving when running it locally 🤔

allouis commented on ghost/core/core/server/services/milestones/service.js:
```diff
@@ -1,4 +1,7 @@
 const DomainEvents = require('@tryghost/domain-events');
+const logging = require('@tryghost/logging');
+
+const JOB_TIMEOUT = 1000 * 60 * 60 * 24 * (Math.floor(Math.random() * 3)); // 0 - 4 days;
```
This is 0-3 days, but the comment says 0-4

aileen replied:
Ah yeah, will update

allouis commented on ghost/core/core/server/services/milestones/service.js:
```diff
@@ -69,10 +72,33 @@ module.exports = {
      * @returns {Promise<object>}
      */
-    async initAndRun() {
+    async initAndRun(customTimeout) {
```
Is this new param only for testing? Perhaps the number should come from config, or we can just use the fake clock in the tests to advance by 4 days 🤷‍♂️

aileen replied:
Yes, this is just for testing purposes

aileen replied:
Added a note. The custom timeout won't last long in the codebase, as we'll soon move to properly schedule the job. I left it in there for now 😅

allouis commented on ghost/core/core/server/services/milestones/service.js:
```diff
@@ -69,10 +72,33 @@ module.exports = {
+
+        logging.info(`Running milestone emails job on ${new Date(jobDate).toString()}`);
+
+        await sleep(timeOut);
```
I don't think we want the init method to wait for up to 3 days, I think it would be better to return from the function and have the code run in a timeout separately

aileen replied:
Switched to use a separate fn which returns a promise with the timeout and doesn't wait for it to be resolved. This should unblock anything that would prevent `boot` from finishing.

allouis reviewed (approved):
Few comments but not major

aileen committed:
Added unblocking `scheduleRun()` fn and updated tests

- We do not want to block the boot process with the previous `sleep()` usage for the timeout and rather return a timeout promise without the need to wait for it to resolve
- Updated days to reflect comment
- Updated tests to check timeout is working
